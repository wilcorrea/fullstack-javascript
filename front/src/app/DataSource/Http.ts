import { AxiosInstance, AxiosRequestConfig } from 'axios'

/**
 * @type {Http}
 */
export default class Http {
  /**
   * @type {string}
   */
  path = '/'

  /**
   * @type {{}}
   */
  static mocks = {}

  /**
   * @type {AxiosInstance}
   */
  private client: AxiosInstance

  /**
   * @type {Http}
   */
  private static single: Http

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static instance (client?: AxiosInstance) {
    if (!this.single) {
      // @ts-ignore
      this.single = new this(client)
    }
    return this.single
  }

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static build (client?: AxiosInstance) {
    // @ts-ignore
    return new this(client)
  }

  /**
   * @param {AxiosInstance} client
   */
  constructor (client: AxiosInstance) {
    this.client = client
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  post (url: string = '', data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.post(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  get (url: string = '', config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.get(url, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  head (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.head(url, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  put (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.put(url, data, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  patch (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.patch(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  delete (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.delete(url, config)
  }

  /**
   * @param {string} path
   * @param value
   */
  static mockRegister (path: string, value: any = null): void {
    // @ts-ignore
    Http.mocks[path] = value
  }

  /**
   * @param {string} path
   * @returns {any}
   */
  static mockRecover (path: string): any {
    // @ts-ignore
    return Http.mocks[path]
  }
}
