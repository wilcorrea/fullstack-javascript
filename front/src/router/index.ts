import Vue from 'vue'
import Router from 'vue-router'
import { redirect } from 'src/app/Util/routing'

import dashboard from 'src/modules/Dashboard/routeFile'
import { updateTitle } from 'src/router/middleware'

Vue.use(Router)

/**
 * @type {string}
 */
export const otherwise = '/dashboard'

/**
 * @type {VueRouter}
 */
const router = new Router()

// after create the router will call the plugin to setup routes and middleware
// the plugin approach allow distribute the router logic to each responsible domain
dashboard(router)

// special route to otherwise
router.addRoutes([redirect('*', otherwise)])

// just a simple middleware
router.afterEach(updateTitle)

export default router
