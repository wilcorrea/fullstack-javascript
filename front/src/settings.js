/**
 * Function to get default app lang
 * @returns {string}
 */
export const i18n = () => env('VUE_APP_I18N', 'pt-br')

/**
 * Casting to wirecard settings
 * Just useful to use Rest (we can use the both in this project =)
 * @returns {{authorization: string, baseURL: string}}
 */
export const wirecard = () => {
  return {
    baseURL: env('VUE_APP_WIRECARD_BASE_URL'),
    authorization: env('VUE_APP_WIRECARD_AUTHORIZATION')
  }
}

/**
 * GraphQL settings to access the backend graphql services
 * @returns {{baseURL: string}}
 */
export const graphQL = () => {
  return { baseURL: env('VUE_APP_GRAPH_QL_BASE_URL') }
}

/**
 * Small helper to access the env
 * ##### I dont know if it is useful, maybe not (it's a WIP) :/ #####
 * @param {string} name
 * @param {*} fallback
 * @returns {string|*}
 */
export const env = (name, fallback = undefined) => {
  try {
    if (process.env[name]) {
      return process.env[name]
    }
  } catch (e) {
    // silent
    console.error(e)
  }
  return fallback
}
