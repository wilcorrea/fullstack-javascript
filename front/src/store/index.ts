import Vue from 'vue'
import Vuex from 'vuex'

Vue.use(Vuex)

/**
 * @interface StoreState
 */
export interface StoreState {
  [key: string]: any
}

/**
 * @interface StoreOptions
 */
export interface StoreOptions {
  state: StoreState,
  mutations: any
}

/**
 * @interface StoreObservable
 */
export interface StoreObservable {
  state: StoreState,
  commit: Function
}

/**
 * Helper to create dynamic stores
 * Useful to reduce boilerplate.. simple and functional
 * Besides the store is standalone and works fine in a lot of places
 * @param {StoreOptions} options
 * @returns {StoreObservable}
 */
export const createStore = (options: StoreOptions): StoreObservable => {
  const { state, mutations } = options
  return {
    state: Vue.observable(state),
    commit (mutation: any, ...args: any) {
      mutations[mutation](state, ...args)
    }
  }
}

/**
 * The root app store!
 * Using createStore to have a simple store to version
 * @type {StoreObservable}
 */
export default createStore({
  // the states of store
  state: {
    version: window.localStorage.getItem('version')
  },
  // the mutations to call with commit
  // ex.: $store.commit('updateVersion')
  mutations: {
    /**
     * @param {StoreState} state
     * @param {string} version
     */
    updateVersion (state: StoreState, version: string) {
      state.version = version
      window.localStorage.setItem('version', version)
    }
  }
})
