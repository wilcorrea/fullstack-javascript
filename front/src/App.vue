<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
