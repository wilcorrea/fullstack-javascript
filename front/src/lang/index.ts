import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
