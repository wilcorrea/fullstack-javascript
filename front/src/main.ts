import Vue from 'vue'

import router from 'src/router'
import store from 'src/store/vuex'
import Boot from 'src/boot'
// import 'src/registerServiceWorker'

import 'material-icons/iconfont/material-icons.css'
import 'src/assets/loading.css'
import 'src/assets/themes/default/@index.styl'

import App from 'src/App.vue'
import 'src/app/Components'

Vue.config.productionTip = false

Vue.use(Boot)

new Vue({
  router,
  store,
  render: h => h(App)
}).$mount('#app')
