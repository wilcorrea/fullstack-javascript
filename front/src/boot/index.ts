import lang from './lang'
import browse from './browse'
import util from './util'
import './polyfill'

/**
 * @type {{install: Function}}
 */
export default {
  install (Vue: any) {
    lang(Vue)
    browse(Vue)
    util(Vue)
  }
}
