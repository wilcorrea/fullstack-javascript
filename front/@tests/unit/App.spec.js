import { shallowMount } from '@vue/test-utils'
import App from 'src/App.vue'
import $store from 'src/store'
import Http from 'src/app/DataSource/Http'

describe('App', () => {
  const RouterView = { template: '<div>RouterView</div>' }

  const wrapper = shallowMount(App, { components: { RouterView } })

  /**
   * @test
   */
  it('check component markup', () => {
    expect(wrapper.html()).toMatch('')
  })

  /**
   * @test
   */
  it('method notify', () => {
    let confirmed = false
    global.confirm = () => true
    global.location.reload = () => (confirmed = true)
    wrapper.vm.notify('jest')
    expect(true).toEqual(wrapper.vm.confirmed)
    expect(true).toEqual(confirmed)

    confirmed = false
    global.confirm = () => (confirmed = true)
    wrapper.vm.confirmed = true
    wrapper.vm.notify('jest')
    expect(false).toEqual(confirmed)

    wrapper.vm.confirmed = false
    confirmed = false
    global.confirm = () => false
    wrapper.vm.notify('jest')
    expect(false).toEqual(wrapper.vm.confirmed)
  })

  /**
   * @test
   */
  it('method receiveVersion', () => {
    $store.state.version = ''
    wrapper.vm.receiveVersion({ data: 'jest' })
    expect('jest').toEqual($store.state.version)

    $store.state.version = 'this.version'
    wrapper.vm.receiveVersion({ data: 'jest' })
    expect('jest').toEqual($store.state.version)
  })

  /**
   * @test
   */
  it('method fetchVersion', async () => {
    $store.state.version = ''
    Http.mockRegister('version', new Promise((resolve) => resolve({ data: 'jest' })))
    const result = await wrapper.vm.fetchVersion()
    expect('jest').toEqual($store.state.version)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
