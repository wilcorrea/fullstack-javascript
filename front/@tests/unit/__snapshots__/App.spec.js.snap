// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render content correctly 1`] = `
<div id="app">
  <router-view-stub></router-view-stub>
</div>
`;
